### ISSUE TITLE
numpy.arange issue with large numbers

### ISSUE THREAD

**Issue by ghobart** [numpy/numpy#18099]
I apologize if this is my bad.
I encountered an issue with numpy.arrange(), I believe that I isolated the problem to the dtype arg giving different results between dtype =np.float and dtype =np.float32. In the case dtype =float32 , numpy.arange  with large start and end values overrided to a step of two.  Here's snippet to demonstrate:

#np_ar_bug.py
import numpy as np
ns= 3635
stline =5682
num_lines=571
n_elements=ns*num_lines

ar_low_val_f32_hack = np.arange(n_elements ,dtype =np.float32) + stline*ns
ar_high_val_f32 = np.arange(  stline*ns,n_elements + stline*ns,1,dtype =np.float32) #overrides to step=2
ar_high_val_f = np.arange(  stline*ns,n_elements + stline*ns,1,dtype =np.float)

print (ar_low_val_f32_hack[:5])
print (ar_high_val_f32[:5])
print (ar_high_val_f[:5])

**Comment by eric-wieser** [comment]
The reason you see this for `np.float32` and not `np.float64` is because of a precision issue. It's worth noting that:
```python
>>> ar_low_val_f32_hack.dtype
dtype('float64')  # not float32
```
If you cast to `f32`, you'll see the precision problem:
```python
>>> np.diff(ar_low_val_f32_hack.astype(np.float32))[:20]
Out[10]:
array([2., 0., 0., 2., 2., 0., 0., 2., 2., 0., 0., 2., 2., 0., 0., 2., 2.,
       0., 0., 2.], dtype=float32)
```
`arange` has some weird logic to try and avoid this type of failure, at the expense of failing in a different way. We ought to document and/or change this, but in your case you probably should just be using a higher precision.

**Comment by ghobart** [comment]
Eric.
Thanks for the quick response. It was a pernicious issue that was difficult
to track down as it only appeared in the later iterations deep within my
process. The script is destined for a supercomputer where the processes
memory footprint  is a critical component hence using f32 instead of f64.
Kind regards.
Geordie

**Comment by ghobart** [comment]
Eric. Thanks again for your quick response. You were much faster than any commercial software vender has ever responded!! Question. It would be very helpful to new users to provide a link to these sort of issues in the reference docs? I am new to numpy and I reviewed this page several times looking for some indication of what was going on with my issue.
https://numpy.org/doc/stable/reference/generated/numpy.arange.html

**Comment by eric-wieser** [comment]
I don't think we want to link to issues from the docs, at least not to explain behavior (explaining history of behavior is a different matter).

The ["arange" github project](https://github.com/numpy/numpy/projects/11#card-52019402) I created to track issue like these. At some point, we should collect together the common issues that people have, and insert them in a "Notes" section for the arange docs you link to.

The docs in question are here: https://github.com/numpy/numpy/blob/31647f1b3e56c2f2a471cd2c3a583311534173f8/numpy/core/_add_newdocs.py#L1337-L1402

A PR to add a notes section that summarizes the issues mentioned in that github "project" would be more than welcome.

**Comment by ghobart** [comment]
Hi Eric.
Thanks for your reply.
I will be discussing contributing with my supervisor in the coming days. I
would like to know the expectations for a respectable PR, so that I can
make an estimate of the commitment of time required.
Can you please point to some introductory information wrt to PRs to give me
some perspective before my discussion with him. Apologies for being such a
nube.
Thanks
Happy new year.
Geordie Hobart

**Comment by eric-wieser** [comment]
There's some information about contributing in [CONTIBUTING.md here](https://github.com/numpy/numpy/blob/31647f1b3e56c2f2a471cd2c3a583311534173f8/.github/CONTRIBUTING.md).

If you don't want to commit to getting up to speed with github and PRs yet though, simply writing a github comment with what you think would make this behavior obvious in the docs would be helpful, and then a maintainer can always make the PR later.
